File: src/main/scala/KafkaToHbase/Kafka2Hbase.java
package KafkaToHbase;

import org.apache.flink.api.common.serialization.SimpleStringSchema;

import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer011;

import java.util.Properties;

/**
 * @author ljz
 */
public class Kafka2Hbase {

    public static void main(String[] args) throws Exception {

        // 引入Flink StreamExecutionEnvironment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // 设置监控数据流时间间隔（官方叫状态与检查点）
        env.enableCheckpointing(1000);

        // 配置kafka和zookeeper的ip和端口
        Properties properties = new Properties();
        properties.setProperty("bootstrap.servers", "140.246.204.127:9092,172.29.0.246:9092");
        properties.setProperty("group.id", "test");


        // 记载kafka和zookeeper的配置
        FlinkKafkaConsumer011 <String> myConsumer = new FlinkKafkaConsumer011 <>("test", new SimpleStringSchema(), properties);

        myConsumer.setStartFromEarliest();     // start from the earliest record possible

//        myConsumer.setStartFromLatest();       // start from the latest record
//        myConsumer.setStartFromTimestamp(...); // start from specified epoch timestamp (milliseconds)
//        myConsumer.setStartFromGroupOffsets(); // the default behaviour
        /*
        // 指定位置
        Map <KafkaTopicPartition, Long> specificStartOffsets = new HashMap <>();
        specificStartOffsets.put(new KafkaTopicPartition("myTopic", 0), 23L);
        myConsumer.setStartFromSpecificOffsets(specificStartOffsets);
*/

//         转换kafka数据类型为flink的dataStream类型
        DataStreamSource stream = env.addSource(myConsumer);

        /* 打印数据内容 */
        stream.print();

        //写入hbase
        stream.addSink(new HBaseWriter());


        // 注册运行名称
        env.execute("the data from kafka");
    }


}

File: src/main/scala/KafkaToHbase/Kafka2hbase.scala
package KafkaToHbase

import java.text.SimpleDateFormat
import java.util.{Date, Properties}

import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer010
import org.apache.hadoop.hbase.client.{ConnectionFactory, Put}
import org.apache.hadoop.hbase.util.Bytes
import org.apache.hadoop.hbase.{HBaseConfiguration, HColumnDescriptor, HTableDescriptor, TableName}
import org.apache.flink.api.scala._


object Kafka2hbase {
  val ZOOKEEPER_URL = "113.125.5.131:2181,113.125.157.152:2181,140.246.131.11:2181,140.246.129.160:2181,140.246.204.127:2181"
  val KAFKA_URL = "140.246.129.160:9092,140.246.204.127:9092"
  val columnFamily = "info"
  val tableName = TableName.valueOf("test")

  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    //设置启动检查点（很重要）
    env.enableCheckpointing(1000)
    // 设置为TimeCharacteristic.EventTime
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    val props = new Properties()
    //    props.setProperty("zookeeper.connect", ZOOKEEPER_URL)
    props.setProperty("bootstrap.servers", KAFKA_URL)
    props.setProperty("group.id", "test")
    val myConsumer = new FlinkKafkaConsumer010[String]("test", new SimpleStringSchema, props)
    myConsumer.setStartFromEarliest()
    val transction: DataStream[String] = env.addSource(myConsumer)
    transction.rebalance.map { value =>
      println(value)
      writeIntoHBase(value)
    }
    env.execute()


  }

  def writeIntoHBase(m: String): Unit = {
    val hbaseconf = HBaseConfiguration.create
    hbaseconf.set("hbase.zookeeper.quorum", ZOOKEEPER_URL)
    hbaseconf.set("hbase.defaults.for.version.skip", "ture")
    val connection = ConnectionFactory.createConnection(hbaseconf)
    val admin = connection.getAdmin

    if (!admin.tableExists(tableName)) {
      admin.createTable(new HTableDescriptor(tableName).addFamily(new HColumnDescriptor(columnFamily)))
    }
    val table = connection.getTable(tableName)
    val df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    val put = new Put(Bytes.toBytes(df.format(new Date())))
    put.addColumn(Bytes.toBytes(columnFamily), Bytes.toBytes("test"), Bytes.toBytes(m))
    table.put(put)
    table.close()

  }

}
